### DISCUSSION TITLE
How to prevent unrelated rerenders?

### DISCUSSION THREAD

**Discussion by tordans** [47ng/nuqs#574] [Q&A] [answered]
I am adding Zustand and Nuqs to a new app.

When adding **Zustand**, I notice local rerenders of all components that consume the getter/setter.
When adding **Nuqs**, I notice global rerenders of all components, no matter if the consume the getter/setter.

I tried abstracting the Nuqs getter/setter into separate as recommended for Zustand in https://tkdodo.eu/blog/working-with-zustand but that does not help. https://github.com/47ng/nuqs/discussions/494#discussioncomment-8435904 suggest that this is not needed ("The state updater function is also referentially stable across renders.") anyways.

But how can I get rid of the unrelated rerenders?

**Video** showing the rerenders  (First no state, then Zustand added, then Nuqs added)

> Update: The app is now open source at https://github.com/tordans/osmcha2
> I switched to using Zustand for all state but the nuqs code is still there and it's just a replacement of ".zustand.ts" to ".nuqs.ts" for the imports.

https://github.com/47ng/nuqs/assets/111561/f3317b5d-0a8a-471d-b3ea-25119be000c9

**Comment by franky47** [comment]
I'd have to look deeper into your example on a big screen next week, but one thing that comes to mind regarding re-renders is the change of behaviour in `useSearchParams` since it became reactive of shallow URL updates.

The documentation mentions that the state updater function is referentially stable, which was relevant before, but now unfortunately any change to the URL will cause **all** `useQueryState(s)` to re-render.

This is not a great experience, I'd have to see if we can implement a scoped `useSearchParam('key')` into Next.js core.

Source: https://x.com/fortysevenfx/status/1796262844062986472

**Reply by tordans** [reply]
@franky47 thanks for confirming the issue. For bigger apps and especially when more state is put in the URL (which is the right way to do it in far more cases that is done today…), this feels like a big issue. So +1 for looking into a solution <3.

Thinking about it, that might also explain why we are having so many rerendering issues with atlas-app in recent weeks … we did update our Next JS version which might have pushed us to a version that always rerenders… ([Demo link](https://radverkehrsatlas.de/regionen/bb-kampagne?v=1&map=11.9/52.399/13.052&config=!(i~bikelanes~a~~sc~!(i~bikelanes~s~!(i~hidden~a~_F)(i~default~a)(i~details~a~_F)(i~width~a~_F))(i~bikelanes*_plus*_presence~s~!(i~default~a~_F))(i~bikelanes*_plus*_width~s~!(i~default~a~_F))(i~bikelanes*_plus*_surface*_smoothness~s~!(i~default~a~_F))(i~bikelanes*_plus*_signs~s~!(i~default~a~_F))(i~bikelanes*_plus*_routes~s~!(i~default~a~_F)))(i~roads~a~~sc~!(i~roads~s~!(i~hidden~a~_F)(i~default~a)(i~sidestreets~a~_F)(i~mainstreets~a~_F))(i~maxspeed~s~!(i~hidden~a)(i~default~a~_F)(i~below30~a~_F)(i~above40~a~_F))(i~roads*_plus*_oneway~s~!(i~default~a~_F))(i~roads*_plus*_footways~s~!(i~default~a~_F)))(i~surface~a~_F~sc~!(i~surfaceRoads~s~!(i~hidden~a~_F)(i~default~a)(i~bad~a~_F))(i~surfaceBikelanes~s~!(i~hidden~a)(i~default~a~_F)(i~bad~a~_F)))(i~boundaries~a~_F~sc~!(i~poiPlaces~s~!(i~hidden~a)(i~default~a~_F)(i~circle~a~_F))(i~poiBoundaries~s~!(i~hidden~a~_F)(i~default~a)(i~category*_district*_label~a~_F)(i~category*_municipality~a~_F)(i~category*_municipality*_label~a~_F))(i~poiPlusBarriers~s~!(i~default~a~_F)))(i~mapillary~a~~sc~!(i~mapillaryCoverage~s~!(i~hidden~a~_F)(i~default~a)(i~all~a~_F)(i~age~a~_F)(i~pano~a~_F)))~&osmNotes=true&data=bb-land-radverkehrsanlagen); but production feels snappy, luckily)

---

Follow up: Does this mean that react compiler (#566) will still help with preventing rerenders due to automatic memoization?

**Reply by roopakv** [reply]
this might be mis-understanding the issue, but if one were to use the native API, would that solve the issue?

https://nextjs.org/docs/app/building-your-application/routing/linking-and-navigating#using-the-native-history-api

**Reply by roopakv** [reply]
never mind, i understand the issue better after reading #516. We have a fairly complicated app and re-rendering every component makes the app not feel as snappy (I'm sure we have some things to fix on our end too). Would be nice if there was a way to avoid this.

**Reply by franky47** [reply]
The best way to work around this is, like every React app, to move your state as down the tree as possible.

You can duplicate calls to the useQueryState(s) hook across components, their state will be kept in sync when updating search params.

**Comment by tordans** [comment]
A bit off topic but related: https://x.com/tkdodo/status/1818338443753136547 is talking about how React Router apparently has the same issue but Tanstack Query does not:
> Another advantage of `@tan_stack` router is that you can subscribe to a subset of query parameters to avoid re-renders when unrelated params change. It makes the url a true state manager like e.g. zustand

**Reply by franky47** [reply, marked as answer]
Note: we're experimenting with "key isolation" that allows scoping re-renders only to the search params that did change, for supported frameworks, in #855 (unfortunately that doesn't include Next.js for which core changes would be necessary).

**Reply by pedro5g** [reply]
Hi, I don't know if this discussion is still open. But recently I had this problem with re-rendering and made a small implementation. So here goes my 2 cents .

Currently, shallow updates (shallow: true) still go through useRouter in Next.js, which causes unnecessary rerenders (see [discussion #574](https://github.com/47ng/nuqs/discussions/574?utm_source=chatgpt.com)
).

I propose separating two modes:

- Client-only mode (shallow: true) → handled internally by the library (via history.pushState/replaceState + pub/sub). No dependency on useRouter.

- Only components subscribed to the affected query param re-render.

- Prevents global rerenders while keeping the URL updated.

- Server mode (shallow: false) → keep using useRouter.push / replace, so Next.js revalidates on the server as usual.

This way:

- We get consistency with Next.js when needed.

- But we also get fine-grained control on the client where rerenders are the main issue.

I built a proof-of-concept implementing this idea, including a QueryManager, an internal queue system, and a rate limiter with Signals:

 https://github.com/pedro5g/trigger.io/tree/main/hooks/query-state

**Reply by franky47** [reply]
> Currently, shallow updates (shallow: true) still go through useRouter in Next.js

In the pages router only, the app router already uses the history API.

Your implementation looks very interesting! It would be nice to run it against the e2e test suite and see how it behaves 👀

**Reply by pedro5g** [reply]
Thanks for the quick reply!

I see, that makes total sense — in the App Router Next.js already uses the History API under the hood, so the issue is mostly visible in the Pages Router.

What I’m exploring with my implementation is a slightly different approach:


- On the client, when shallow: true, updates are applied directly via the History API + a pub/sub manager (without relying on useRouter).



- On the server, or when a full navigation is required, it falls back to router.push/replace.


This brings a few benefits:


- Granular control: developers can choose when to sync only the client state vs. when to involve the server.

- Less re-renders: components only re-render when the subscribed key changes (pub/sub style, similar to Zustand).

- Rate-limiting built-in: updates are scheduled and debounced with a signal-based queue to avoid race conditions.

I also read in https://github.com/47ng/nuqs/pull/855 about the experiment with key isolation — scoping re-renders only to the search params that actually changed.
From what I understood, you mentioned that this isn’t currently possible in Next.js because it would require core changes.

In my approach, since I’m not relying on useRouter for shallow: true, but instead patching the History API directly with a pub/sub mechanism, I can get something very close to key isolation (each hook instance only re-renders when its own param changes).

I’d be happy to try running it against the e2e suite of nuqs to check compatibility!

**Reply by franky47** [reply]
I use a similar approach for the React Router & Remix implementation, where the History API is used with a cache layer (essentially a subscription, but without a central store, something I need to look into, see #1099) for shallow: true, but fallback to the router APIs for calling server loaders.

The pattern is kind of foreign and not really recommended in the "idiomatic React Router / Remix way", but it does yield good performance improvements (that folks in those communities had been asking for).

One drawback of this approach is that the stock APIs for those routers don't follow external updates to the History API, so if you were to have components that listen to the URL via the useRouter().query or useSearchParams/useLocation, they'd get out of sync with the URL.
In the React Router / Remix adapters, nuqs exposes a read-only `useOptimisticSearchParams` hook that is reactive to shallow updates (but uses unparsed string to serve as a replacement for the useSearchParams hook).

**Reply by pedro5g** [reply]
That drawback you mentioned about the router APIs not following external updates to the History API is something I also noticed when experimenting.

In my implementation, though, I tried to mitigate this issue: with the monkey patch, whenever useQueryState is called, it automatically triggers a recalculation of useSearchParams. This way, the sync between the History API and the router hooks is maintained, and the usual “out of sync” problem doesn’t really occur.

Since nuqs already provides stronger control and parsing of query strings than the native hooks, this extra synchronization step makes the experience more consistent in practice.

Do you see any risks or downsides with this approach?

<img width="1156" height="588" alt="Captura de tela 2025-09-18 194831" src="https://github.com/user-attachments/assets/7395f406-002e-4f90-9649-ac7602a63487" />


[Gravando 2025-09-18 155903.webm](https://github.com/user-attachments/assets/87019849-62a6-4eb5-af7c-5b94e5fdf81b)

**Reply by franky47** [reply]
> whenever useQueryState is called, it automatically triggers a recalculation of useSearchParams.

How did you achieve that? And does it not negate the fine grained subscriptions / key isolation you have in place with your useQueryState (as components relying on the router core hooks should re-render no matter what change happened in the URL, since you don't know what they are going to consume).

SInce your implementation is a fork+edit of nuqs (a lot of the code is very similar), would you be interested in contributing some of your architectural choices back into nuqs?

**Reply by pedro5g** [reply]
Regarding your first question: I achieved the recalculation of useSearchParams by monkey-patching the router methods (push/replace). This way, the updates go through a global queryManager, which centralizes the state changes. Since the monkey-patch also updates the history API (pushState/replaceState), both the Next.js internals and the browser history remain in sync.

Because of this setup, everything stays perfectly synchronized
the global queryManager ensures fine-grained subscriptions (only the components subscribed to a specific query key re-render),
while useSearchParams and router-level consumers also get updated automatically

As for the concern about losing isolation: in practice, the granularity is preserved since components depending on useQueryState still only update when their key changes. The synchronization layer doesn’t broadcast unnecessary updates; it just makes sure all sources of truth (Next.js router, browser history, and queryManager) are consistent.

And yes, I’d definitely be interested in contributing some of these architectural choices back into nuqs!

**Reply by pedro5g** [reply]
One last thing I’d like to ask.
Given the current architecture of nuqs, what would you consider the safest and easiest way to integrate my approach (monkey-patched router + global queryManager for fine-grained subscriptions) into the existing codebase?

My goal is to keep the isolation guarantees already present in nuqs, while making shallow updates smoother by handling them client-side first. Do you think this could fit behind an adapter layer (like the current router-specific implementations, or would you see it more as an experimental mode/flag that could be toggled on?

I’d love your advice here, since I want to make sure any contribution aligns with the long-term direction of nuqs

**Reply by colsondonohue** [reply]
Hello! Any updates on this @franky47 @pedro5g?

**Reply by franky47** [reply]
I've had @pedro5g's code on my downloads folder for a while but I've unfortunately been lacking time to properly look at it, sorry about that.

Aside from key isolation in Next.js, the single store approach has a lot of other benefits, but would be a big rewrite. I'm hoping to free a bit more time to work on that in the coming months.

**Reply by pedro5g** [reply]
Hi! 
If you're interested, I actually have some free time next week and would be happy to help push this forward. I recently built a reduced version of the approach for personal use, and intercepting changes at the native browser history API level worked surprisingly well. It gave me good results in terms of performance and framework agnosticism.
In React/Next.js specifically, the strategy allowed me to keep all existing features while removing the need for providers entirely  the only requirement was a small wrapper around the router.
If this direction aligns with what you're envisioning for nuqs, I’d be glad to collaborate on it.

**Comment by pedro5g** [comment]
And @colsondonohue 
if you're interested, I can also share the reduced version of my nuqs implementation so you can use and test it directly.

**Reply by colsondonohue** [reply]
would definitely be interested. does it maintain the same interface, etc.?

**Reply by pedro5g** [reply]
You can take a look here and test it 
https://www.npmjs.com/package/@query-state/next

**Reply by colsondonohue** [reply]
cool, will take a look! would also love a jotai-style `set` hook, where I can use a typed setter without subscribing to changes

**Reply by pedro5g** [reply]
That's cool, Zustand and Jotai was my inspiration for state management. Plus, it has a refined auto type parse.
